tests/Paysera/WalletApi/MapperTest.php

```diff
@@ -3,41 +3,720 @@
+                'expectedException' => Paysera_WalletApi_Exception_LogicException::class
+            ],
+            'no description' => [
+                'project' => $project3,
+                'expectedOutput' => [
+                    'title' => 'Test project',
+                    'wallet_id' => 12345,
+                ],
+                'expectedException' => null
+            ],
+            'no wallet_id' => [
+                'project' => $project4,
+                'expectedOutput' => [
+                    'title' => 'Test project',
+                    'description' => 'Test description',
+                ],
+                'expectedException' => null
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider decodeProjectDataProvider
+     */
+    public function testDecodeProject($data, $expectedOutput)
+    {
+        $result = $this->mapper->decodeProject($data);
+
+        if ($expectedOutput !== null) {
+            $this->assertEquals($expectedOutput->getId(), $result->getId());
+            $this->assertEquals($expectedOutput->getTitle(), $result->getTitle());
+            $this->assertEquals($expectedOutput->getDescription(), $result->getDescription());
+            $this->assertEquals($expectedOutput->getWalletId(), $result->getWalletId());
+        }
+    }
+
+    public function decodeProjectDataProvider()
+    {
+        $project1 = new Paysera_WalletApi_Entity_Project();
+        $project1->setId(1);
+        $project1->setTitle("Test project");
+        $project1->setDescription("Test description");
+        $project1->setWalletId(12345);
+
+        $project2 = new Paysera_WalletApi_Entity_Project();
+        $project2->setId(2);
+        $project2->setTitle("Test project");
+        $project2->setWalletId(12345);
+
+        $project3 = new Paysera_WalletApi_Entity_Project();
+        $project3->setId(3);
+        $project3->setTitle("Test project");
+        $project3->setDescription("Test description");
+
+        return [
+            'all data' => [
+                'data' => [
+                    'id' => 1,
+                    'title' => 'Test project',
+                    'description' => 'Test description',
+                    'wallet_id' => 12345,
+                ],
+                'expectedOutput' => $project1,
+            ],
+            'empty description' => [
+                'data' => [
+                    'id' => 2,
+                    'title' => 'Test project',
+                    'wallet_id' => 12345,
+                ],
+                'expectedOutput' => $project2,
+            ],
+            'empty wallet_id' => [
+                'data' => [
+                    'id' => 3,
+                    'title' => 'Test project',
+                    'description' => 'Test description',
+                ],
+                'expectedOutput' => $project3,
+            ],
+        ];
+    }
+    public function testDecodePaymentAllData()
+    {
+        $data = [
```

Comment:
@PianistaMichal Please, move this code to DataProvider

Reply:
I added there separate test instead of dataProvider because of different data to check. It's better to use here separate test